### DISCUSSION TITLE
Comparison with Xamarin and projects with similar goals

### DISCUSSION THREAD

**Discussion by SoonDead** [AvaloniaUI/Avalonia#5479] [Q&A] [answered]
I have no experience in cross platform .NET development, I have only tried Xamarin for a 2-3 screen demo app running on Android iOS and WP. Still it seems the .NET world is opening, some things are alpha, some are stable (yet work like some alpha), but it's becoming harder and harder to explore all options.

How is this different from Xamarin.Forms for example? Why would someone choose one and not the other?

I think it would be useful to see a document about the most shiny selling points of this framework.

**Comment by jkoritzinsky** [comment, marked as answer]
Xamarin.Forms uses the native controls on each platform to render instead of drawing customizable controls. Additionally, Xamarin.Forms only works on iOS, Android, WP, and UWP.

Conversely, Avalonia works on Windows, Linux, iOS, Android (may have regressed), UWP (missing glue code though), Mac OS X, and is very easily extensible to embedded platforms that support mono.

Xamarin.Forms is more stable currently. If you're doing mobile then Xamarin.Forms would be more useful right now, but Avalonia will give you more flexibility. However, once we stabilize Android, then Avalonia will be the only UI toolkit that works across both mobile and desktop platforms.

**Comment by JimSEOW** [comment]
Xamarin Forms supports the following OSs
iOS, MacOS
Android
GTK (Linux)
WPF (soon x86/x64)
UWP ( W10, W10M, W10A, IoT Core, IoT Core Pro)

**Comment by weitzhandler** [comment]
@JimSEOW I'm not aware of XF support for WPF, it does support UWP, though not sure WPF. Can you please include any reference for that fact?

**Comment by JimSEOW** [comment]
@weitzhandler 
It was announced in Build 2017,[ it is expected to come Q3 2017](https://github.com/xamarin/Xamarin.Forms/pull/895#issuecomment-313441039)
https://blog.xamarin.com/glimpse-future-xamarin-forms-3-0/

**Comment by JimSEOW** [comment]
@weitzhandler  It is important that the different cross platform WPF efforts learn from each other and help each other!

**Comment by juepiezhongren** [comment]
we want ava and xf mixed together

**Comment by jmacato** [comment]
@juepiezhongren that would be pointless, since the two projects are divergent in their goals (Xamarin.Forms focuses with native experience in their target platforms, while Avalonia is for devs that want their apps to have uniform UI/UX across all supported platforms). Besides the untold amount of manpower required to merge the two projects or have them interoperate would be better focused on other aspects of the project that will be tangibly more beneficial to all.

**Comment by kevmoens** [comment]
> @JimSEOW I'm not aware of XF support for WPF, it does support UWP, though not sure WPF. Can you please include any reference for that fact?
To work with Xamarin Forms with WPF you need to add a new project to your solution for WPF and add the corresponding Xamarin Forms Nuget
https://www.nuget.org/packages/Xamarin.Forms.Platform.WPF/

**Comment by ToolmakerSteve** [comment]
The essential difference is "**Xamarin.Forms uses the native controls on each platform**".

XF Pros:
* UI that looks and acts like the platform UI - because it is the platform UI's widgets.
* Rich UI interaction and page navigation model. (Instead of re-working in each platform's UI paradigm, solve this once for your app.)
* Backed by Microsoft. As MAUI, becoming part of ".Net Everywhere".
* Developed for mobile devices first - so those targets are well-supported,

XF Cons (compared to writing native UI yourself on mobile devices; I haven't yet written tests to compare Avalonia to XF):
* MUCH harder for XF devs to get this 100% reliable and high performance. As of today, there are [**3,212 open issues for XF**](https://github.com/xamarin/Xamarin.Forms/issues). OTOH, that's a lot of app devs finding any issues in it, a mixed blessing.
* Larger memory footprint.
* Slower layout/measuring cycle than prepping native controls directly. (Takes longer before a complex page appears, compared to writing native UI yourself.)

I've spent a LOT of time over the past several years pushing XF to the limit. I'm a hardcore .Net fan. I love the XF concept. (I actively help people surmount XF problems on StackOverflow.)

I'm skeptical that MAUI will ever be FULLY competitive (in memory resources and first-time-page-appearance latency and glitch-free reliability) with writing directly in each platform's native UI. I do believe it will get to the point where it is a worthwhile tradeoff, saving substantial development time vs extra resources etc. But this will take time.

**Avalonia takes a different approach: pixel-perfect control over the look of your app**. Give it a screen rectangle on any device, and it will render to it. This is inherently a much simple problem than what XF does. Logically, the result is leaner, and easier to make robust.

Once ".Net for Android" and "Net for iOS" are finalized, XF will no longer have an inherent advantage in running well on those platforms. Any alternative solution, including Avalonia, will gain full advantage of Microsoft's maintenance of Miguel de Icaza's awesome work in making C# run well on mobile platforms.

I'm currently experimenting with Avalonia, in part to avoid XF's problems. For situations where "pixel-perfect control" is wanted, instead of "use each platform's widgets". Prototyping high-performance layered 2D and 3D on Windows today; when ".net for iOS/Android" mature, will include those platforms also.

**Comment by robloo** [comment]
@SoonDead Also see the below link for a high level comparison of the differences and pros/cons. There is a lot to consider and this is intended to help make strategic decisions. It also explains why some of the differences in architecture exist. (Note that Maui is the same as Xamarin.Forms in this document)

https://github.com/robloo/PublicDocs/blob/master/XAMLFrameworkComparison.md

**Reply by npolyak** [reply]
Good review!

**Comment by npolyak** [comment]
In my experience Xamarin is too limited by the native implementations, while Avalonia is not. Xamarin primitives are buttons, dropdowns etc, while in Avalonia buttons and comboboxes and other controls can be composed out of smaller primitives - borders, shapes - etc. 
Besides Xamarin is too mobile oriented - even though they say that they are going to support desktop - I have not quite seen it yet. 
Also Xamarin would produce a different set of dll files for each platform, while under Avalonia it is the same dlls and executables that will run on every platform.
